tyrian/js/src/main/scala/tyrian/Cmd.scala

```diff
@@ -1,67 +1,76 @@
-      observable: Task.Observable[Err, Success],
-      toMessage: Either[Err, Success] => Msg
-  ) extends Cmd[Msg]:
-    def map[OtherMsg](f: Msg => OtherMsg): Run[Err, Success, OtherMsg] = Run(observable, toMessage andThen f)
-    def attempt[OtherMsg](resultToMessage: Either[Err, Success] => OtherMsg): Run[Err, Success, OtherMsg] =
-      Run(observable, resultToMessage)
+  /** Represents runnable concurrent task that produces a message */
+  final case class Run[F[_], A, Msg](
+      task: F[A],
+      toMsg: A => Msg
+  ) extends Cmd[F, Msg]:
+    def map[OtherMsg](f: Msg => OtherMsg): Run[F, A, OtherMsg] = Run(task, toMsg andThen f)
   object Run:
+    @targetName("Cmd.Run separate param lists")
+    def apply[F[_], A, Msg](task: F[A])(toMessage: A => Msg): Run[F, A, Msg] =
+      Run(task, toMessage)
 
-    def apply[Err, Success, Msg](toMessage: Either[Err, Success] => Msg)(
-        observable: Task.Observable[Err, Success]
-    ): Run[Err, Success, Msg] =
-      Run(observable, toMessage)
-
-    final case class ImcompleteRunCmd[Err, Success](observable: Task.Observable[Err, Success]):
-      def attempt[Msg](resultToMessage: Either[Err, Success] => Msg): Run[Err, Success, Msg] =
-        Run(observable, resultToMessage)
-
-    def apply[Err, Success](observable: Task.Observable[Err, Success]): ImcompleteRunCmd[Err, Success] =
-      ImcompleteRunCmd(observable)
-
-  final case class RunTask[Err, Success, Msg](task: Task[Err, Success], toMessage: Either[Err, Success] => Msg)
-      extends Cmd[Msg]:
-    def map[OtherMsg](f: Msg => OtherMsg): RunTask[Err, Success, OtherMsg] = RunTask(task, toMessage andThen f)
+  /** Merge two commands into a single one */
+  case class Combine[F[_], Msg](cmd1: Cmd[F, Msg], cmd2: Cmd[F, Msg]) extends Cmd[F, Msg]:
+    def map[OtherMsg](f: Msg => OtherMsg): Combine[F, OtherMsg] = Combine(cmd1.map(f), cmd2.map(f))
 
-  case class Combine[Msg](cmd1: Cmd[Msg], cmd2: Cmd[Msg]) extends Cmd[Msg]:
-    def map[OtherMsg](f: Msg => OtherMsg): Combine[OtherMsg] = Combine(cmd1.map(f), cmd2.map(f))
-
-  case class Batch[Msg](cmds: List[Cmd[Msg]]) extends Cmd[Msg]:
-    def map[OtherMsg](f: Msg => OtherMsg): Batch[OtherMsg] = this.copy(cmds = cmds.map(_.map(f)))
+  /** Treat many commands as one */
+  case class Batch[F[_], Msg](cmds: List[Cmd[F, Msg]]) extends Cmd[F, Msg]:
+    def map[OtherMsg](f: Msg => OtherMsg): Batch[F, OtherMsg] = this.copy(cmds = cmds.map(_.map(f)))
   object Batch:
-    def apply[Msg](cmds: Cmd[Msg]*): Batch[Msg] =
+    def apply[F[_], Msg](cmds: Cmd[F, Msg]*): Batch[F, Msg] =
       Batch(cmds.toList)
+
+  given [F[_], Msg]: Monoid[Cmd[F, Msg]] = new Monoid[Cmd[F, Msg]] {
+    def empty: Cmd[F, Msg]                                   = Cmd.Empty
+    def combine(a: Cmd[F, Msg], b: Cmd[F, Msg]): Cmd[F, Msg] = Cmd.merge(a, b)
+  }
+
+  def combineAll[F[_], A](list: List[Cmd[F, A]])(implicit M: Monoid[Cmd[F, A]]): Cmd[F, A] =
+    list.foldRight(M.empty)(M.combine)
```

Comment:
Interesting: I merged it, and went in to apply the same fix to `Sub`s (2 minute lunchtime job) - but this actually breaks the tests. Looks like `Monoid[A].combineAll` does a `foldLeft` rather than a `foldRight`. Which means that instead of this:

```scala
Cmd.Combine(
  Cmd.Emit(10),
  Cmd.Combine(
    Cmd.Emit(20),
    Cmd.Emit(30)
  )
)
```

You get this:

```scala
Cmd.Combine(
  Cmd.Combine(
    Cmd.Emit(10),
    Cmd.Emit(20)
  ),
  Cmd.Emit(30)
)
```

It amounts to the same thing in the end, but less of a harmless refactor than I anticipated.

Reply:
Fixed and pushed - thanks!

Reply:
Right, those have to be the same thing or else it violates the Monoid associativity law  `(a |+| b) |+| c == a |+| (b |+| c)`.

Running Cats discipline tests would also catch this :)